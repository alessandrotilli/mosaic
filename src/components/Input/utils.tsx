import React from "react";
import { InputAdornment as MUIInputAdornment } from "@material-ui/core";

import { IInputAdornment } from "../../types/Input";
import IconButton from "../IconButton";

export const getAdornment = (adornment?: IInputAdornment) => {
  if (!adornment) {
    return;
  }

  const { icon, onClick } = adornment;
  return (
    <MUIInputAdornment position="end">
      <IconButton disabled={!onClick} icon={icon} onClick={onClick!} size="small" />
    </MUIInputAdornment>
  );
};
