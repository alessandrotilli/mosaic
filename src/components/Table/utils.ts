export const COLUMN_CHECKBOX_PATH = "checkbox-selection";
export const COLUMN_ROW_ACTIONS_PATH = "row-actions";

export const TOOLBAR_HEIGHT = 64;
export const TOOLBAR_HEIGHT_MOBILE = 56;

export const PAGINATION_TOOLBAR_HEIGHT = 64;
