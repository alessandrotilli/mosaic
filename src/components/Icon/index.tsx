import React, { cloneElement, FC, PropsWithChildren, ReactElement, useCallback, useMemo } from "react";
import { ClassNames } from "@emotion/react";
import { keyframes, Skeleton as MUISkeleton } from "@mui/material";

import { IIcon, IIconDimensions, IRenderedIcon } from "../../types/Icon";
import { logWarn } from "../../utils/logger";
import Adornment from "../Adornment";

import { iconsCatalog } from "./utils";

export const DATA_CY_DEFAULT = "icon";

const ICON_DIMENSIONS: IIconDimensions = {
  small: 20,
  medium: 24,
  large: 35,
};

const Icon: FC<PropsWithChildren<IIcon>> = ({
  badge,
  children,
  dataCy = DATA_CY_DEFAULT,
  forwarded = {},
  loading = false,
  name,
  rotate = false,
  size = "medium",
  style: externalStyle,
  tooltip,
}) => {
  const dimensions = useMemo(() => {
    const dimension = ICON_DIMENSIONS[size];
    return { height: dimension, width: dimension };
  }, [size]);

  const props = useMemo(() => {
    let props: IRenderedIcon = {
      ...forwarded,
      "data-cy": dataCy,
      fontSize: size,
    };

    if (externalStyle) {
      props = { ...props, style: { ...externalStyle } };
    }

    if (children) {
      props = { ...props, style: { ...props.style, ...dimensions } };
    }

    return props;
  }, [children, dataCy, dimensions, externalStyle, forwarded, size]);

  const rotateKeyframes = useMemo(
    () => keyframes`
      from: {
        transform: "rotate(0deg)";
      }
      to: {
        transform: "rotate(360deg)";
      }
  `,
    []
  );

  const renderWithAnimation = useCallback(
    (element: any) => {
      return (
        <ClassNames>
          {({ css }) => {
            let animation = "none";
            if (rotate) {
              animation = css`
                ${rotateKeyframes} 2s linear infinite
              `;
            }

            let className = css`
              animation: ${animation};
            `;

            if (forwarded.className) {
              className = forwarded.className;
            }

            return (
              <Adornment badge={badge} tooltip={tooltip}>
                {cloneElement(element as ReactElement<any>, { ...props, className })}
              </Adornment>
            );
          }}
        </ClassNames>
      );
    },
    [badge, forwarded, props, rotate, rotateKeyframes, tooltip]
  );

  if (loading) {
    return <MUISkeleton variant="rectangular" style={{ ...dimensions }} />;
  }

  if (children) {
    return renderWithAnimation(children);
  }

  if (!name) {
    logWarn("Icon", "Skip rendering, both children and name are not set");
    return null;
  }

  const icon = iconsCatalog[name];
  return renderWithAnimation(icon);
};

export default Icon;
