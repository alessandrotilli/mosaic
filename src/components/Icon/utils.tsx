import React from "react";
import MUIAccountCircleIcon from "@material-ui/icons/AccountCircle";
import MUIAddIcon from "@material-ui/icons/Add";
import MUIAppsIcon from "@material-ui/icons/Apps";
import MUIArrowUpwardIcon from "@material-ui/icons/ArrowUpward";
import MUIBlockIcon from "@material-ui/icons/Block";
import MUIBrightness4Icon from "@material-ui/icons/Brightness4";
import MUIBusinessIcon from "@material-ui/icons/Business";
import MUICheckIcon from "@material-ui/icons/Check";
import MUICheckBoxIcon from "@material-ui/icons/CheckBox";
import MUICheckBoxOutlineBlankIcon from "@material-ui/icons/CheckBoxOutlineBlank";
import MUIChevronLeftIcon from "@material-ui/icons/ChevronLeft";
import MUIChevronRightIcon from "@material-ui/icons/ChevronRight";
import MUICloseIcon from "@material-ui/icons/Close";
import MUICropIcon from "@material-ui/icons/Crop";
import MUIDeleteIcon from "@material-ui/icons/Delete";
import MUIDetailsIcon from "@material-ui/icons/Details";
import MUIDraftsIcon from "@material-ui/icons/Drafts";
import MUIEditIcon from "@material-ui/icons/Edit";
import MUIErrorIcon from "@material-ui/icons/Error";
import MUIFilterListIcon from "@material-ui/icons/FilterList";
import MUIFirstPageIcon from "@material-ui/icons/FirstPage";
import MUIGroupWorkIcon from "@material-ui/icons/GroupWork";
import MUIHomeIcon from "@material-ui/icons/Home";
import MUIKeyboardArrowDownIcon from "@material-ui/icons/KeyboardArrowDown";
import MUIKeyboardArrowUpIcon from "@material-ui/icons/KeyboardArrowUp";
import MUILanguageIcon from "@material-ui/icons/Language";
import MUILastPageIcon from "@material-ui/icons/LastPage";
import MUIListIcon from "@material-ui/icons/List";
import MUIMailIcon from "@material-ui/icons/Mail";
import MUIMenuIcon from "@material-ui/icons/Menu";
import MUINotificationsIcon from "@material-ui/icons/Notifications";
import MUIOpenInNew from "@material-ui/icons/OpenInNew";
import MUIPauseIcon from "@material-ui/icons/Pause";
import MUIPlayArrowIcon from "@material-ui/icons/PlayArrow";
import MUIPlaylistAddCheckIcon from "@material-ui/icons/PlaylistAddCheck";
import MUIPowerSettingsNewIcon from "@material-ui/icons/PowerSettingsNew";
import MUIPrintIcon from "@material-ui/icons/Print";
import MUIRefreshIcon from "@material-ui/icons/Refresh";
import MUISaveIcon from "@material-ui/icons/Save";
import MUIScheduleIcon from "@material-ui/icons/Schedule";
import MUISearchIcon from "@material-ui/icons/Search";
import MUISendIcon from "@material-ui/icons/Send";
import MUISettingsIcon from "@material-ui/icons/Settings";
import MUISignalWifi1BarOutlinedIcon from "@material-ui/icons/SignalWifi1BarOutlined";
import MUITimelineIcon from "@material-ui/icons/Timeline";
import MUIVisibilityIcon from "@material-ui/icons/Visibility";
import MUIVisibilityOffIcon from "@material-ui/icons/VisibilityOff";
import MUIWarningIcon from "@material-ui/icons/Warning";
import MUIWifiOffOutlinedIcon from "@material-ui/icons/WifiOffOutlined";
import MUIWifiOutlinedIcon from "@material-ui/icons/WifiOutlined";

import { Icons, IIconsCatalog } from "../../types/Icon";

export const iconsCatalog: IIconsCatalog = {
  [Icons.account]: <MUIAccountCircleIcon />,
  [Icons.add]: <MUIAddIcon />,
  [Icons.apps]: <MUIAppsIcon />,
  [Icons.block]: <MUIBlockIcon />,
  [Icons.brightness]: <MUIBrightness4Icon />,
  [Icons.business]: <MUIBusinessIcon />,
  [Icons.chart_timeline]: <MUITimelineIcon />,
  [Icons.check]: <MUICheckIcon />,
  [Icons.checkbox]: <MUICheckBoxIcon />,
  [Icons.checkbox_empty]: <MUICheckBoxOutlineBlankIcon />,
  [Icons.clock]: <MUIScheduleIcon />,
  [Icons.close]: <MUICloseIcon />,
  [Icons.connection_slow]: <MUISignalWifi1BarOutlinedIcon />,
  [Icons.crop]: <MUICropIcon />,
  [Icons.delete]: <MUIDeleteIcon />,
  [Icons.details]: <MUIDetailsIcon />,
  [Icons.down]: <MUIKeyboardArrowDownIcon />,
  [Icons.edit]: <MUIEditIcon />,
  [Icons.error]: <MUIErrorIcon />,
  [Icons.filter]: <MUIFilterListIcon />,
  [Icons.first]: <MUIFirstPageIcon />,
  [Icons.group]: <MUIGroupWorkIcon />,
  [Icons.home]: <MUIHomeIcon />,
  [Icons.language]: <MUILanguageIcon />,
  [Icons.last]: <MUILastPageIcon />,
  [Icons.left]: <MUIChevronLeftIcon />,
  [Icons.list_check]: <MUIPlaylistAddCheckIcon />,
  [Icons.list]: <MUIListIcon />,
  [Icons.mail_open]: <MUIDraftsIcon />,
  [Icons.mail]: <MUIMailIcon />,
  [Icons.menu]: <MUIMenuIcon />,
  [Icons.notifications]: <MUINotificationsIcon />,
  [Icons.open_new]: <MUIOpenInNew />,
  [Icons.pause]: <MUIPauseIcon />,
  [Icons.play]: <MUIPlayArrowIcon />,
  [Icons.power]: <MUIPowerSettingsNewIcon />,
  [Icons.print]: <MUIPrintIcon />,
  [Icons.refresh]: <MUIRefreshIcon />,
  [Icons.right]: <MUIChevronRightIcon />,
  [Icons.save]: <MUISaveIcon />,
  [Icons.search]: <MUISearchIcon />,
  [Icons.send]: <MUISendIcon />,
  [Icons.settings]: <MUISettingsIcon />,
  [Icons.up_arrow]: <MUIArrowUpwardIcon />,
  [Icons.up]: <MUIKeyboardArrowUpIcon />,
  [Icons.visibility]: <MUIVisibilityIcon />,
  [Icons.visibility_off]: <MUIVisibilityOffIcon />,
  [Icons.warn]: <MUIWarningIcon />,
  [Icons.wifi_off]: <MUIWifiOffOutlinedIcon />,
  [Icons.wifi_on]: <MUIWifiOutlinedIcon />,
};
