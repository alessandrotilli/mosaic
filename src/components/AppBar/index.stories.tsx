import React from "react";
import { action } from "@storybook/addon-actions";
import { boolean, select, text } from "@storybook/addon-knobs";

import { Icons } from "../../types/Icon";
import { getAllComposedDataCy } from "../../utils";
import IntlProviderMock, { LocaleMock, MessageMock } from "../../utils/mocks/IntlProviderMock";
import { getDocumentationPage } from "../../utils/stories";

import AppBar, { AppBarWithProps, DATA_CY_DEFAULT, DATA_CY_SHORTCUT, LOCALIZABLE_PROPS, SUBPARTS_MAP } from ".";

export default {
  title: "AppBar",
  component: AppBarWithProps,
  parameters: {
    ...getDocumentationPage({
      basedOn: "@material-ui/core/AppBar",
      component: "AppBar",
      e2eTestInfo: {
        dataCyDefault: DATA_CY_DEFAULT,
        dataCyShortcut: DATA_CY_SHORTCUT,
        subpartsSuffixes: getAllComposedDataCy(SUBPARTS_MAP),
      },
      localizableProps: LOCALIZABLE_PROPS,
    }),
  },
};

export const Canvas = () => (
  <AppBar
    actions={[{ icon: Icons.mail, onClick: action("On Mail Click") }]}
    dataCy={text("dataCy", DATA_CY_DEFAULT)}
    localized={boolean("localized", false)}
    menu={{ icon: Icons.home, onClick: action("On Menu Click") }}
    onTitleClick={action("On Title Click")}
    title={text("title", "AppBar Title")}
    userMenu={[{ label: "Logout", onClick: action("On Logout") }]}
    username={text("username", "mosaic")}
  />
);

export const CustomStyle = () => (
  <AppBar
    actions={[
      { icon: Icons.language, onClick: () => {}, style: { backgroundColor: "yellow", color: "black" } },
      { icon: Icons.notifications, onClick: () => {}, style: { backgroundColor: "green", color: "white" } },
    ]}
    style={{ backgroundColor: "red" }}
    title="AppBar"
    userMenu={[{ label: "Logout", onClick: () => {} }]}
  />
);

export const Localized = () => (
  // IntlProviderMock simulates external IntlProvider context
  <IntlProviderMock locale={select("locale", LocaleMock, LocaleMock.en)}>
    <AppBar localized title={MessageMock.title} userMenu={[{ label: MessageMock.button, onClick: action("Button") }]} />
  </IntlProviderMock>
);

export const WithUsername = () => (
  <AppBar
    title="Mosaic"
    userMenu={[
      { label: "View Details", onClick: action("View Details") },
      { label: "Logout", onClick: action("Logout") },
    ]}
    username="mosaic@github.com"
  />
);
