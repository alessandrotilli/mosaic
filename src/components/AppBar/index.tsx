import React, { FC } from "react";
import { useMemo } from "react";
import { AppBar as MUIAppBar, Toolbar as MUIToolbar, useTheme } from "@material-ui/core";

import { IAppBar } from "../../types/AppBar";
import { Icons } from "../../types/Icon";
import { IMenu } from "../../types/Menu";
import { TypographyVariants } from "../../types/Typography";
import { getComposedDataCy, suppressEvent } from "../../utils";
import localized, { ILocalizableProperty } from "../../utils/hocs/localized";
import { logWarn } from "../../utils/logger";
import IconButton from "../IconButton";
import Menu from "../Menu";
import Typography from "../Typography";

export const DATA_CY_DEFAULT = "appbar";
export const DATA_CY_SHORTCUT = "title";
export const LOCALIZABLE_PROPS: ILocalizableProperty[] = [
  { name: DATA_CY_SHORTCUT, type: "string" },
  { name: "userMenu.label", type: "any[]" },
  // TODO: review for localized hoc, adding "user.items.label"
];

export const SUBPARTS_MAP = {
  action: {
    label: "Action (at index n)",
    value: (n = `{n}`) => `action-${n}`,
  },
  localesMenu: {
    label: "Locales Menu",
  },
  localesMenuItem: {
    label: "Locales Menu Item (at index n)",
    value: (n = `{n}`) => `locales-menu-item-${n}`,
  },
  menu: {
    label: "Menu",
  },
  userMenu: {
    label: "User Menu",
  },
  userMenuItem: {
    label: "User Menu Item (at index n)",
    value: (n = `{n}`) => `user-menu-item-${n}`,
  },
  titleClickable: {
    label: "Title Clickable",
  },
  titleText: {
    label: "Title Text",
  },
};

const AppBar: FC<IAppBar> = ({
  actions = [],
  dataCy = DATA_CY_DEFAULT,
  locale,
  menu,
  onTitleClick,
  style,
  title,
  user,
  userMenu = [],
  username,
}) => {
  const theme = useTheme();

  const userProps = useMemo((): IMenu | undefined => {
    if (user) {
      return { ...user };
    }

    if (!userMenu || !userMenu.length) {
      return;
    }

    logWarn(
      "AppBar",
      'Deprecation notice, "userMenu" and "username" will be deprecated in favour of "user". Please update your code.'
    );

    const userMenuItems = userMenu.map((menuItem) => ({ ...menuItem, value: menuItem.label }));
    return {
      items: userMenuItems,
      label: username,
    };
  }, [user, userMenu, username]);

  return (
    <MUIAppBar data-cy={dataCy} position="sticky" style={style}>
      <MUIToolbar style={{ alignItems: "center", display: "flex", justifyContent: "space-between" }}>
        <div style={{ alignItems: "center", display: "flex" }}>
          {menu && (
            <IconButton dataCy={getComposedDataCy(dataCy, SUBPARTS_MAP.menu)} icon={menu.icon} onClick={menu.onClick} />
          )}
          {title && (
            <div
              data-cy={getComposedDataCy(dataCy, SUBPARTS_MAP.titleClickable)}
              onClick={(event) => {
                suppressEvent(event);
                onTitleClick && onTitleClick();
              }}
              style={{
                borderRadius: `${theme.shape.borderRadius}px`,
                cursor: onTitleClick ? "pointer" : "default",
                padding: `${theme.spacing(0.5)}px ${theme.spacing(1)}px`,
                userSelect: "none",
              }}
            >
              <Typography dataCy={getComposedDataCy(dataCy, SUBPARTS_MAP.titleText)} variant={TypographyVariants.title}>
                {title}
              </Typography>
            </div>
          )}
        </div>
        <div style={{ alignItems: "center", display: "flex" }}>
          {actions.map(({ icon, onClick, style }, index) => (
            <IconButton
              key={`action-${index}`}
              dataCy={getComposedDataCy(dataCy, SUBPARTS_MAP.action, index)}
              icon={icon}
              onClick={onClick}
              style={{ marginRight: `${theme.spacing(0.5)}px`, ...style }}
            />
          ))}
          {locale && (
            <Menu dataCy={getComposedDataCy(dataCy, SUBPARTS_MAP.localesMenu)} icon={Icons.language} {...locale} />
          )}
          {userProps && (
            <Menu
              dataCy={getComposedDataCy(dataCy, SUBPARTS_MAP.userMenu)}
              icon={Icons.account}
              style={{ textTransform: "lowercase" }}
              {...userProps}
            />
          )}
        </div>
      </MUIToolbar>
    </MUIAppBar>
  );
};

export const AppBarWithProps = AppBar;

export default localized(AppBar, {
  dataCyShortcut: DATA_CY_SHORTCUT,
  localizableProps: LOCALIZABLE_PROPS,
});
