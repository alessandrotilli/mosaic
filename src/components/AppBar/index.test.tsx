import React from "react";
import renderer from "react-test-renderer";

import { IAppBar } from "../../types/AppBar";
import { Icons } from "../../types/Icon";
import { getComposedDataCy } from "../../utils";
import { getLocalizedMessageMock, MessageMock } from "../../utils/mocks/LocaleMock";
import { getTestableComponent, IPartialTestOptions, ITestOptions } from "../../utils/tests";

import AppBar, { DATA_CY_DEFAULT, SUBPARTS_MAP } from ".";

const DEFAULT_TEST_OPTIONS: ITestOptions<IAppBar> = {
  dataCy: DATA_CY_DEFAULT,
  domNode: "header",
  localized: true,
  props: {},
};

const getAppBarTestable = (options?: IPartialTestOptions<IAppBar>) =>
  getTestableComponent(AppBar, DEFAULT_TEST_OPTIONS, options);

describe("AppBar test suite:", () => {
  it("default", () => {
    const { element, wrapper } = getAppBarTestable();
    expect(wrapper).toHaveLength(1);

    const snapshotWrapper = renderer.create(element).toJSON();
    expect(snapshotWrapper).toMatchSnapshot();
  });

  it("dataCy", () => {
    const dataCy = "custom";
    const { element, wrapper } = getAppBarTestable({ dataCy, props: { dataCy } });
    expect(wrapper).toHaveLength(1);

    const snapshotWrapper = renderer.create(element).toJSON();
    expect(snapshotWrapper).toMatchSnapshot();
  });

  it("localized", () => {
    const label = MessageMock.confirm;
    const title = MessageMock.title;
    const { wrapper } = getAppBarTestable({
      dataCy: title,
      props: { localized: true, title, userMenu: [{ label, onClick: jest.fn() }] },
    });

    const titleElementDataCy = getComposedDataCy(title, SUBPARTS_MAP.titleText);
    const titleElement = wrapper.find(`h2[data-cy='${titleElementDataCy}']`);
    expect(titleElement.text()).toEqual(getLocalizedMessageMock(title, "en"));

    const userMenuEntryDataCy = getComposedDataCy(title, SUBPARTS_MAP.userMenuItem, 0);
    const userMenuEntry = wrapper.find(`li[data-cy='${userMenuEntryDataCy}']`);
    expect(userMenuEntry.text()).toEqual(getLocalizedMessageMock(label, "en"));
  });

  it("actions", () => {
    const onClick = jest.fn();
    const { element, wrapper } = getAppBarTestable({ props: { actions: [{ icon: Icons.account, onClick }] } });

    const actionButtonDataCy = getComposedDataCy(DATA_CY_DEFAULT, SUBPARTS_MAP.action, 0);
    const actionButton = wrapper.find(`button[data-cy='${actionButtonDataCy}']`);
    actionButton.simulate("click");
    expect(onClick).toHaveBeenCalledTimes(1);

    const actionButtonIconDataCy = `${actionButtonDataCy}-icon`;
    const actionButtonIcon = actionButton.find(`Icon[dataCy='${actionButtonIconDataCy}']`);
    expect(actionButtonIcon.prop("name")).toEqual(Icons.account);

    const snapshotWrapper = renderer.create(element).toJSON();
    expect(snapshotWrapper).toMatchSnapshot();
  });

  it("custom content", () => {
    const customText = "Custom Content";
    const { element, wrapper } = getAppBarTestable({
      props: { children: <div id="content">{customText}</div> } as any,
    });

    const customContent = wrapper.find("#content");
    expect(customContent.text()).toEqual(customText);

    const snapshotWrapper = renderer.create(element).toJSON();
    expect(snapshotWrapper).toMatchSnapshot();
  });

  it("locale", () => {
    const onItemClick = jest.fn();
    const { wrapper } = getAppBarTestable({
      props: {
        locale: {
          items: [
            { label: "English", value: "en" },
            { label: "Italian", value: "it" },
          ],
          label: "en",
          onItemClick,
        },
      },
    });

    const localeButtonDataCy = getComposedDataCy(DATA_CY_DEFAULT, SUBPARTS_MAP.localesMenu);
    const localeButton = wrapper.find(`button[data-cy='${localeButtonDataCy}']`);
    const localeButtonIconDataCy = `${localeButtonDataCy}-icon`;
    const localeButtonIcon = localeButton.find(`Icon[dataCy='${localeButtonIconDataCy}']`);
    expect(localeButtonIcon.prop("name")).toEqual(Icons.language);
    localeButton.simulate("click");

    const localeEntryDataCy = getComposedDataCy(DATA_CY_DEFAULT, SUBPARTS_MAP.localesMenuItem, 1);
    const localeEntry = wrapper.find(`li[data-cy='${localeEntryDataCy}']`);
    expect(localeEntry.text()).toEqual("Italian");
    localeEntry.simulate("click");
    expect(onItemClick).toHaveBeenCalledTimes(1);
    expect(onItemClick).toHaveBeenCalledWith("it");
    localeButton.simulate("blur");
  });

  it("menu", () => {
    const onClick = jest.fn();
    const { element, wrapper } = getAppBarTestable({ props: { menu: { icon: Icons.account, onClick } } });

    const menuButtonDataCy = getComposedDataCy(DATA_CY_DEFAULT, SUBPARTS_MAP.menu);
    const menuButton = wrapper.find(`button[data-cy='${menuButtonDataCy}']`);
    menuButton.simulate("click");
    expect(onClick).toHaveBeenCalledTimes(1);

    const menuButtonIconDataCy = `${menuButtonDataCy}-icon`;
    const menuButtonIcon = menuButton.find(`Icon[dataCy='${menuButtonIconDataCy}']`);
    expect(menuButtonIcon.prop("name")).toEqual(Icons.account);

    const snapshotWrapper = renderer.create(element).toJSON();
    expect(snapshotWrapper).toMatchSnapshot();
  });

  it("title", () => {
    const onTitleClick = jest.fn();
    const { element, wrapper } = getAppBarTestable({ props: { onTitleClick, title: "Title" } });

    const titleWrapperDataCy = getComposedDataCy(DATA_CY_DEFAULT, SUBPARTS_MAP.titleClickable);
    const titleWrapper = wrapper.find(`div[data-cy='${titleWrapperDataCy}']`);
    titleWrapper.simulate("click");
    expect(onTitleClick).toHaveBeenCalledTimes(1);

    const titleTextDataCy = getComposedDataCy(DATA_CY_DEFAULT, SUBPARTS_MAP.titleText);
    const titleElement = wrapper.find(`h2[data-cy='${titleTextDataCy}']`);
    expect(titleElement.text()).toEqual("Title");

    const snapshotWrapper = renderer.create(element).toJSON();
    expect(snapshotWrapper).toMatchSnapshot();
  });

  it("userMenu", () => {
    const onClick = jest.fn();
    const { wrapper } = getAppBarTestable({
      props: { user: { items: [{ label: "Logout", onClick, value: "logout" }], label: "User" } },
    });

    const userMenuButtonDataCy = getComposedDataCy(DATA_CY_DEFAULT, SUBPARTS_MAP.userMenu);
    const userMenuButton = wrapper.find(`button[data-cy='${userMenuButtonDataCy}']`);
    const userMenuButtonIconDataCy = `${userMenuButtonDataCy}-icon`;
    const userMenuButtonIcon = userMenuButton.find(`Icon[dataCy='${userMenuButtonIconDataCy}']`);
    expect(userMenuButtonIcon.prop("name")).toEqual(Icons.account);
    userMenuButton.simulate("click");

    const userMenuEntryDataCy = getComposedDataCy(DATA_CY_DEFAULT, SUBPARTS_MAP.userMenuItem, 0);
    const userMenuEntry = wrapper.find(`li[data-cy='${userMenuEntryDataCy}']`);
    expect(userMenuEntry.text()).toEqual("Logout");
    userMenuEntry.simulate("click");
    expect(onClick).toHaveBeenCalledTimes(1);
    userMenuButton.simulate("blur");

    // TODO#lb: add coverage for setUserMenuAnchor(null) when closing userMenu
    // <MUIMenu
    //   id={`${dataCy}-user-menu`}
    //   anchorEl={userMenuAnchor}
    //   anchorOrigin={MENU_ITEMS_ANCHORING}
    //   keepMounted
    //   transformOrigin={MENU_ITEMS_ANCHORING}
    //   open={!!userMenuAnchor}
    //   onClose={() => setUserMenuAnchor(null)}
    // >
  });

  it("username", () => {
    const onClick = jest.fn();
    const username: string = "mos@ic";
    const { wrapper } = getAppBarTestable({
      props: { user: { items: [{ label: "Logout", onClick, value: "logout" }], label: username } },
    });

    const userMenuButtonDataCy = getComposedDataCy(DATA_CY_DEFAULT, SUBPARTS_MAP.userMenu);
    const userMenuButton = wrapper.find(`button[data-cy='${userMenuButtonDataCy}']`);
    const userMenuButtonLabel = userMenuButton.find(".MuiButton-label");
    expect(userMenuButtonLabel.text()).toEqual(username);
  });
});
