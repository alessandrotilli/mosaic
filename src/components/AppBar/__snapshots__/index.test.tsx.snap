// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar test suite: actions 1`] = `
<header
  className="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary MuiPaper-elevation4"
  data-cy="appbar"
>
  <div
    className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "space-between",
      }
    }
  >
    <div
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "flex": 1,
        }
      }
    />
    <div
      style={
        {
          "alignItems": "center",
          "display": "flex",
        }
      }
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
        data-cy="appbar-action-0"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          {
            "marginRight": "4px",
          }
        }
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            data-cy="appbar-action-0-icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</header>
`;

exports[`AppBar test suite: custom content 1`] = `
<header
  className="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary MuiPaper-elevation4"
  data-cy="appbar"
>
  <div
    className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "space-between",
      }
    }
  >
    <div
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "flex": 1,
        }
      }
    >
      <div
        id="content"
      >
        Custom Content
      </div>
    </div>
    <div
      style={
        {
          "alignItems": "center",
          "display": "flex",
        }
      }
    />
  </div>
</header>
`;

exports[`AppBar test suite: dataCy 1`] = `
<header
  className="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary MuiPaper-elevation4"
  data-cy="custom"
>
  <div
    className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "space-between",
      }
    }
  >
    <div
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "flex": 1,
        }
      }
    />
    <div
      style={
        {
          "alignItems": "center",
          "display": "flex",
        }
      }
    />
  </div>
</header>
`;

exports[`AppBar test suite: default 1`] = `
<header
  className="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary MuiPaper-elevation4"
  data-cy="appbar"
>
  <div
    className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "space-between",
      }
    }
  >
    <div
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "flex": 1,
        }
      }
    />
    <div
      style={
        {
          "alignItems": "center",
          "display": "flex",
        }
      }
    />
  </div>
</header>
`;

exports[`AppBar test suite: menu 1`] = `
<header
  className="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary MuiPaper-elevation4"
  data-cy="appbar"
>
  <div
    className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "space-between",
      }
    }
  >
    <div
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "flex": 1,
        }
      }
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
        data-cy="appbar-menu"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            data-cy="appbar-menu-icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      style={
        {
          "alignItems": "center",
          "display": "flex",
        }
      }
    />
  </div>
</header>
`;

exports[`AppBar test suite: title 1`] = `
<header
  className="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary MuiPaper-elevation4"
  data-cy="appbar"
>
  <div
    className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "space-between",
      }
    }
  >
    <div
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "flex": 1,
        }
      }
    >
      <div
        data-cy="appbar-title-clickable"
        onClick={[Function]}
        style={
          {
            "borderRadius": "4px",
            "cursor": "pointer",
            "padding": "4px 8px",
            "userSelect": "none",
          }
        }
      >
        <h2
          className="MuiTypography-root MuiTypography-h6"
          data-cy="appbar-title-text"
        >
          Title
        </h2>
      </div>
    </div>
    <div
      style={
        {
          "alignItems": "center",
          "display": "flex",
        }
      }
    />
  </div>
</header>
`;
